ipa-core/src/helpers/buffers/ordering_sender.rs

```diff
@@ -79,59 +66,43 @@ impl State {
     // sent will be the same size, so we settle for a less strict check that should at least
     // prevent reaching a deadlock.
     fn write<M: Message>(&mut self, m: &M, cx: &Context<'_>) -> Poll<()> {
-        debug_assert!(
-            self.spare != 0 || self.buf.capacity() % M::Size::USIZE == 0,
-            "invalid spare capacity for OrderingSender (see docs)",
-        );
-
-        if !self.accept_writes() {
+        if !self.buf.can_write() {
             Self::save_waker(&mut self.write_ready, cx);
             return Poll::Pending;
         }
 
-        let b = &mut self.buf[self.written..];
-        assert!(
-            M::Size::USIZE <= b.len(),
-            "expect message size {:?} to fit in available buffer; only {:?} of {:?} available",
-            M::Size::USIZE,
-            self.buf.capacity() - self.written,
-            self.buf.capacity(),
-        );
-        self.written += M::Size::USIZE;
-        m.serialize(GenericArray::from_mut_slice(&mut b[..M::Size::USIZE]));
-
-        if !self.accept_writes() {
+        self.buf.next().write(m);
+
+        // todo: test tht it is not in_full call
+        if self.buf.can_read() {
             Self::wake(&mut self.stream_ready);
         }
+
         Poll::Ready(())
     }
 
     fn take(&mut self, cx: &Context<'_>) -> Poll<Vec<u8>> {
-        if self.written > 0 && (self.written + self.spare >= self.buf.len() || self.closed) {
-            let v = self.buf[..self.written].to_vec();
-            self.written = 0;
+        if self.buf.can_read() {
+            if !self.buf.can_write() {
+                // We are ready to unblock writers by taking some data that we know is there off
+                // the buffer
+                Self::wake(&mut self.write_ready);
+            }
 
-            Self::wake(&mut self.write_ready);
-            Poll::Ready(v)
+            Poll::Ready(self.buf.take())
```

Comment:
And include a test that reveals this problem.

Reply:
we discussed it in our meeting - it is not easy to reproduce out of order problem because of the lock around the `State` object that serializes access to internals. Even if we wake another thread before we take the data out, it will still block until the lock is released and that only happens after `.take()` call.

I am going to change the order anyway because if we ever to remove the lock which `CircularBuf` can make possible, then this will cause issues. And I'll try to write a test that unfortunately won't fail today, but it could fail if there is no lock and wrong order